### PR TITLE
Feature - Multiple customers per user (cg/cgweb)

### PR DESCRIPTION
**Pull request by patrikgrenfeldt**
See https://github.com/Clinical-Genomics/cgweb/pull/300

### CODE REVIEW

File: cg/cli/get.py [hunk 1]
```diff
@@ -109,7 +109,9 @@ def family(
         if customer_obj is None:
             LOG.error(f"{customer}: customer not found")
             context.abort()
-        families = context.obj["status_db"].families(customer=customer_obj, enquiry=family_ids[-1])
+        families = context.obj["status_db"].families(
```
Comment (moonso) [inline cg/cli/get.py:112]:
```suggestion
        cases: List[models.Family] = context.obj["status_db"].families(
```

File: cg/server/api.py [hunk 2]
```diff
@@ -74,29 +74,13 @@ def submit_order(order_type):
...
     except HTTPError as error:
-        return abort(make_response(jsonify(message=error), 401))
+        return abort(make_response(jsonify(message=str(error)), 401))
 
     return jsonify(
         project=result["project"], records=[record.to_dict() for record in result["records"]]
     )
 
 
-@BLUEPRINT.route("/customers")
```
Comment (moonso) [inline cg/server/api.py:84]: why remove these?

File: cg/server/api.py [hunk 3]
```diff
@@ -113,24 +97,25 @@ def families():
         records = db.cases_to_mip_analyze()
         count = len(records)
     else:
-        customer_obj = None if g.current_user.is_admin else g.current_user.customer
+        customer_objs = None if g.current_user.is_admin else g.current_user.customers
```
Comment (moonso) [inline cg/server/api.py:100]:
```suggestion
        customer_objs: Optional[List[models.Customer]] = None if g.current_user.is_admin else g.current_user.customers
```

Comment (moonso) [reply cg/server/api.py:100]: I like the type hints since it makes one think if stuff can be done in other ways :)

File: cg/server/api.py [hunk 4]
```diff
@@ -113,24 +97,25 @@ def families():
...
         case_query = db.families(
             enquiry=request.args.get("enquiry"),
-            customer=customer_obj,
+            customers=customer_objs,
             action=request.args.get("action"),
         )
         count = case_query.count()
         records = case_query.limit(30)
+
     data = [case_obj.to_dict(links=True) for case_obj in records]
```
Comment (moonso) [inline cg/server/api.py:109]: mmm data, don't we just love it! At least we could specify a little bit what the data is:
```suggestion
    data: List[dict] = [case_obj.to_dict(links=True) for case_obj in records]
```

File: cg/server/api.py [hunk 5]
```diff
@@ -113,24 +97,25 @@ def families():
...
     return jsonify(families=data, total=count)
 
 
 @BLUEPRINT.route("/families_in_customer_group")
 def families_in_customer_group():
     """Fetch families in customer_group."""
-    customer_obj = None if g.current_user.is_admin else g.current_user.customer
+    customer_objs = None if g.current_user.is_admin else g.current_user.customers
```
Comment (moonso) [inline cg/server/api.py:116]: type hint

File: cg/server/api.py [hunk 6]
```diff
@@ -113,24 +97,25 @@ def families():
...
     families_q = db.families_in_customer_group(
```
Comment (moonso) [inline cg/server/api.py:117]: type hint and perhaps a more descriptive variable name, `family_query_result` perhaps?

File: cg/server/api.py [hunk 7]
```diff
@@ -113,24 +97,25 @@ def families():
...
-        enquiry=request.args.get("enquiry"), customer=customer_obj
+        enquiry=request.args.get("enquiry"), customers=customer_objs
     )
     count = families_q.count()
```
Comment (moonso) [inline cg/server/api.py:120]: type hint

File: cg/server/api.py [hunk 8]
```diff
@@ -113,24 +97,25 @@ def families():
...
     records = families_q.limit(30)
```
Comment (moonso) [inline cg/server/api.py:121]: type hint

File: cg/server/api.py [hunk 9]
```diff
@@ -144,7 +129,7 @@ def family(family_id):
     case_obj = db.family(family_id)
```
Comment (moonso) [inline cg/server/api.py:129]:
```suggestion
    case_obj: Optional[models.Family] = db.family(family_id)
```

File: cg/server/api.py [hunk 10]
```diff
@@ -144,7 +129,7 @@ def family(family_id):
...
     if case_obj is None:
         return abort(404)
-    if not g.current_user.is_admin and (g.current_user.customer != case_obj.customer):
+    if not g.current_user.is_admin and (case_obj.customer not in g.current_user.customers):
         return abort(401)
 
     data = case_obj.to_dict(links=True, analyses=True)
```
Comment (moonso) [inline cg/server/api.py:135]:
```suggestion
    data: dict = case_obj.to_dict(links=True, analyses=True)
```

File: cg/server/api.py [hunk 11]
```diff
@@ -178,8 +164,8 @@ def samples():
     elif request.args.get("status") == "sequencing":
         samples_q = db.samples_to_sequence()
     else:
-        customer_obj = None if g.current_user.is_admin else g.current_user.customer
-        samples_q = db.samples(enquiry=request.args.get("enquiry"), customer=customer_obj)
+        customer_objs = None if g.current_user.is_admin else g.current_user.customers
```
Comment (moonso) [inline cg/server/api.py:167]: type hint

File: cg/server/api.py [hunk 12]
```diff
@@ -178,8 +164,8 @@ def samples():
...
+        samples_q = db.samples(enquiry=request.args.get("enquiry"), customers=customer_objs)
```
Comment (moonso) [inline cg/server/api.py:168]: type hint and more descriptive variable name

File: cg/server/api.py [hunk 13]
```diff
@@ -188,9 +174,9 @@ def samples():
 @BLUEPRINT.route("/samples_in_customer_group")
 def samples_in_customer_group():
     """Fetch samples in a customer group."""
-    customer_obj = None if g.current_user.is_admin else g.current_user.customer
+    customer_objs = None if g.current_user.is_admin else g.current_user.customers
```
Comment (moonso) [inline cg/server/api.py:177]: type hint

File: cg/server/api.py [hunk 14]
```diff
@@ -188,9 +174,9 @@ def samples():
...
     samples_q = db.samples_in_customer_group(
```
Comment (moonso) [inline cg/server/api.py:178]: same

File: cg/server/api.py [hunk 15]
```diff
@@ -226,8 +213,8 @@ def sample_in_customer_group(sample_id):
 @BLUEPRINT.route("/pools")
 def pools():
     """Fetch pools."""
-    customer_obj = None if g.current_user.is_admin else g.current_user.customer
-    pools_q = db.pools(customer=customer_obj, enquiry=request.args.get("enquiry"))
+    customer_objs = None if g.current_user.is_admin else g.current_user.customers
+    pools_q = db.pools(customers=customer_objs, enquiry=request.args.get("enquiry"))
```
Comment (moonso) [inline cg/server/api.py:217]: same

File: cg/server/api.py [hunk 16]
```diff
@@ -276,9 +263,8 @@ def analyses():
 @BLUEPRINT.route("/options")
 def options():
     """Fetch various options."""
-    customer_objs = (
-        db.Customer.query.all() if g.current_user.is_admin else [g.current_user.customer]
-    )
+    customer_objs = db.Customer.query.all() if g.current_user.is_admin else g.current_user.customers
```
Comment (moonso) [inline cg/server/api.py:266]: type hint

File: cg/store/api/findbusinessdata.py [hunk 17]
```diff
@@ -87,17 +93,19 @@ def families(
         )
 
         records = records.filter_by(action=action) if action else records
-
         return records.order_by(models.Family.created_at.desc())
 
     def families_in_customer_group(
-        self, *, customer: models.Customer = None, enquiry: str = None
+        self, *, customers: [models.Customer] = None, enquiry: str = None
```
Comment (moonso) [inline cg/store/api/findbusinessdata.py:99]:
```suggestion
        self, *, customers: List[models.Customer] = None, enquiry: str = None
```

File: cg/store/api/findbusinessdata.py [hunk 18]
```diff
@@ -211,10 +212,15 @@ def new_invoice_id(self) -> int:
         ids = [inv.id for inv in query]
         return max(ids) + 1 if ids else 0
 
-    def pools(self, *, customer: models.Customer, enquiry: str = None) -> Query:
+    def pools(self, *, customers: [models.Customer] = None, enquiry: str = None) -> Query:
```
Comment (moonso) [inline cg/store/api/findbusinessdata.py:215]:
```suggestion
    def pools(self, *, customers: Optional[List[models.Customer]] = None, enquiry: str = None) -> Query:
```

File: cg/store/api/findbusinessdata.py [hunk 19]
```diff
@@ -234,9 +240,15 @@ def sample(self, internal_id: str) -> models.Sample:
         """Fetch a sample by lims id."""
         return self.Sample.query.filter_by(internal_id=internal_id).first()
 
-    def samples(self, *, customer: models.Customer = None, enquiry: str = None) -> Query:
+    def samples(self, *, customers: [models.Customer] = None, enquiry: str = None) -> Query:
```
Comment (moonso) [inline cg/store/api/findbusinessdata.py:243]:
```suggestion
    def samples(self, *, customers: Optional[List[[models.Customer]] = None, enquiry: str = None) -> Query:
```

File: cg/store/api/findbusinessdata.py [hunk 20]
```diff
@@ -259,14 +271,17 @@ def samples_by_ids(self, **identifiers) -> Query:
         return records.order_by(models.Sample.internal_id.asc())
 
     def samples_in_customer_group(
-        self, *, customer: models.Customer = None, enquiry: str = None
+        self, *, customers: [models.Customer] = None, enquiry: str = None
```
Comment (moonso) [inline cg/store/api/findbusinessdata.py:274]:
```suggestion
        self, *, customers: Optional[List[models.Customer]] = None, enquiry: str = None
```

**Review by moonso — CHANGES_REQUESTED** [review]
Cool!

File: cg/cli/get.py (hunk 1 above)
Comment (patrikgrenfeldt) [reply cg/cli/get.py:112]: fixed

File: cg/server/api.py (hunk 3 above)
Comment (patrikgrenfeldt) [reply cg/server/api.py:100]: good point, fixed

File: cg/server/api.py (hunk 4 above)
Comment (patrikgrenfeldt) [reply cg/server/api.py:109]: fixed

File: cg/server/api.py (hunk 5 above)
Comment (patrikgrenfeldt) [reply cg/server/api.py:116]: fixed

File: cg/server/api.py (hunk 16 above)
Comment (patrikgrenfeldt) [reply cg/server/api.py:266]: fixed

File: cg/server/api.py (hunk 2 above)
Comment (moonso) [reply cg/server/api.py:84]: I think we should keep!

**Comment by moonso** [comment]
Nice work @patrikgrenfeldt !

**Review by moonso — APPROVED** [review]

**Comment by patrikgrenfeldt** [comment]
Deployed on Hasta:
image

Upgraded db:
image

Deployed on clinical-db:
image

**Pull request merged**